### PR TITLE
Improve HMR handling for styles, persisted islands

### LINKED ISSUE #3445: 🐛 BUG: svelte styles are not loaded on HMR
### What version of `astro` are you using?

1.0.0-beta.33

### Are you using an SSR adapter? If so, which one?

None

### What package manager are you using?

npm

### What operating system are you using?

Linux

#### To Reproduce
- Create a new project a project using the `npm create astro@1.0.0-beta.33`
- Choose the "Just the basics" template
- Choose to add the Svelte integration only
- Create a simple svelte component with styles, add it to index page
- Trigger HRM

#### Description
  When HRM is triggered - even if the styles did not change, or if the change is not in a svelte file - all styles from svelte components stop showing. After a page refresh it loads normally, but vanishes as soon as the next HMR hits.
  
  Other styles, like the ones written on `.astro` files, behave normally.
  
  The issue stated on #3219 is also happening here, but it doesn't seem to be the same thing - this issue is also happening on Chrome.

### Link to Minimal Reproducible Example

https://stackblitz.com/edit/github-nxhq4u?file=src/components/Counter.svelte

### Participation

- [ ] I am willing to submit a pull request for this issue.

### PR DESCRIPTION
**Pull request by natemoo-re**
Closes #3445

## Changes

- Improves Astro style HMR by adopting the same inline style technique Vite uses.
- Improves island hydration during HMR
- Fixes a Svelte-specific CSS HMR bug (#3445)
- Updates a few of the E2E test fixtures that were pegged to a specific version of `astro`

## Testing

E2E tests updated with a check to make sure that component styles aren't lost after saving `.astro` files.  This mainly tests the Svelte-specific bug

There's always opportunities for more/better test coverage here!  Unfortunately I didn't find a way to reliably test the flash of unstyled content bug this PR was targeted at.  Playwright makes sure the page is finished loading before letting our tests run, that means the CSS flash isn't really visible to the test logic at all.

## Docs

N/A

## Demos

[Framework components](https://www.loom.com/share/6ee02825915c494d9d575db3132941e7)
[Tailwind CSS](https://www.loom.com/share/30e3b42e7b24486fb7ade8db0d2c5857)
[Nested framework components](https://www.loom.com/share/9cb2b343dbfa47a89f7c95db2b3239ab)

### CODE REVIEW

**Comment by tony-sull** [comment]
Looks like there's a few corner cases to track down here first, but this one has me a little nervous that there may be hidden regressions here.

- Haven't tracked it down fully yet, but I'm actually getting both the inline style and the build style `<link>`

I see it flash with the `<link>` including `localhost:3000` in the URL, not sure yet where that's coming from but that seems to be why the `<link>` isn't being removed during HMR as expected here

In general, I think it'd be ideal if `astro dev` always used inline styles (similar to Vite) but there are a few challenges there and I'm not sure how feasible that is for now

File: packages/astro/src/runtime/client/hmr.ts
Comment (tony-sull) [inline packages/astro/src/runtime/client/hmr.ts:25]: This is the fix for the Svelte component bug

Any `<style>` blocks that were injected by Astro or Vite as part of HMR should be persisted during the DOM diffing update

File: packages/astro/src/runtime/client/hmr.ts
Comment (tony-sull) [inline packages/astro/src/runtime/client/hmr.ts:16]: Makes sure that micromorph keeps the `astro-root` elements in the DOM as-is

File: packages/astro/src/runtime/client/hmr.ts
Comment (tony-sull) [inline packages/astro/src/runtime/client/hmr.ts:51]: For .astro styles, they are now injected as inline `<style>` blocks (similar to what Vite does internally, including with Svelte styles)

File: packages/astro/src/runtime/client/hmr.ts
Comment (tony-sull) [inline packages/astro/src/runtime/client/hmr.ts:90]: Keeps track of style blocks by file ID to make sure they are replaced on the next HMR update (instead of duplicating the styles)

**Review by natemoo-re — COMMENTED** [review]
I can't approve this PR since I opened it, but LGTM!

**Comment by tony-sull** [comment]
I want to do a bit more testing here before considering it ready to merge, but the last two changes make a big improvement to HMR :eyes:

In `dev`, most styles are now inlined as `<style>` blocks which will support HMR without a flash of unstyled content **and** avoid fighting with Vite's own HMR handlers that collide with our own HMR handler

**Comment by AllanChain** [comment]
I can confirm that #3370 and #3219 are also fully solved in this PR.

File: packages/astro/src/core/render/dev/css.ts
Comment (tony-sull) [inline packages/astro/src/core/render/dev/css.ts:72]: A few use cases still build styles to JS modules rather than plain CSS strings, the main one I came across being CSS modules

File: packages/astro/e2e/fixtures/client-only/package.json
Comment (matthewp) [inline packages/astro/e2e/fixtures/client-only/package.json:6]: Will all of these changes not breaking people using `npm init astro`?

Comment (matthewp) [reply packages/astro/e2e/fixtures/client-only/package.json:6]: Oh, nm, these are e2e tests, not the examples 🤦‍♂️

**Review by matthewp — APPROVED** [review]

**Comment by natemoo-re** [comment]
@felixsanz thanks! Responded on Discord.

**Pull request merged**
